libsolutil/Dominator.h

```diff
@@ -0,0 +1,366 @@
+			});
+		};
+		dfs(_entry, dfs);
+
+		// Process the vertices in decreasing order of the DFS number
+		for (DfsIndex wIdx: m_verticesInDFSOrder | ranges::views::reverse | ranges::views::transform(toDfsIndex))
+		{
+			// step 3
+			// NOTE: this is an optimization, i.e. performing the step 3 before step 2.
+			// The goal is to process the bucket in the beginning of the loop for the vertex ``w``
+			// instead of ``parent[w]`` in the end of the loop as described in the original paper.
+			// Inverting those steps ensures that a bucket is only processed once and
+			// it does not need to be erased.
+			// The optimization proposal is available here: https://jgaa.info/index.php/jgaa/article/view/paper119/2847 pg.77
+			for (DfsIndex vIdx: bucket[wIdx])
+			{
+				DfsIndex uIdx = eval(vIdx);
+				solAssert(uIdx <= vIdx);
+				idom[vIdx] = (semi[uIdx] < semi[vIdx]) ? uIdx : wIdx;
+			}
+
+			// step 2
+			for (DfsIndex vIdx: predecessors[wIdx])
+			{
+				DfsIndex uIdx = eval(vIdx);
+				solAssert(uIdx <= vIdx);
+				if (semi[uIdx] < semi[wIdx])
+					semi[wIdx] = semi[uIdx];
+			}
+			bucket[semi[wIdx]].emplace_back(wIdx);
+			link(parent[wIdx], wIdx);
+			solAssert(ancestor[wIdx] == parent[wIdx]);
+		}
+
+		// step 4
+		// Compute idom in dfs order
+		idom[0] = 0;
+		for (DfsIndex wIdx: m_verticesInDFSOrder | ranges::views::drop(1) | ranges::views::transform(toDfsIndex))
+			if (idom[wIdx] != semi[wIdx])
+				idom[wIdx] = idom[idom[wIdx]];
+
+		return idom;
+	}
+
+	/// Path compression updates the ancestors of vertices along
+	/// the path to the ancestor with the minimum label value.
+	void compressPath(
+		std::vector<DfsIndex>& _ancestor,
+		std::vector<DfsIndex>& _label,
+		std::vector<DfsIndex>& _semi,
+		DfsIndex _vIdx
+	) const
+	{
+		solAssert(_ancestor[_vIdx] != std::numeric_limits<size_t>::max());
+		DfsIndex uIdx = _ancestor[_vIdx];
+		if (_ancestor[uIdx] != std::numeric_limits<size_t>::max())
+		{
+			compressPath(_ancestor, _label, _semi, uIdx);
+			if (_semi[_label[uIdx]] < _semi[_label[_vIdx]])
+				_label[_vIdx] = _label[uIdx];
+			_ancestor[_vIdx] = _ancestor[uIdx];
+		}
+		solAssert(_label[uIdx] <= _label[_vIdx]);
+	}
+
+	/// Build dominator tree from the immediate dominators set.
+	/// The function groups all the vertices ids that are immediately dominated by a vertex.
+	void buildDominatorTree()
+	{
+		// m_immediateDominators is guaranteed to have at least one element after findingDominators() is executed.
+		solAssert(m_immediateDominators.size() > 0);
+		solAssert(m_immediateDominators.size() == m_verticesInDFSOrder.size());
+		solAssert(m_immediateDominators[0] == 0);
+
+		// Ignoring the entry node since no one dominates it.
+		for (DfsIndex index = 1; index < m_verticesInDFSOrder.size(); ++index)
+		{
+			VId vId = m_verticesInDFSOrder[index];
+			solAssert(m_dfsIndexByVertexId.count(vId));
+			solAssert(index == m_dfsIndexByVertexId.at(vId));
```

Comment:
Are you assuming that the number of vertices / the graph you're coming the dominators over is a single connected component? This assertion may fail otherwise if, f.ex., you have a graph like `A <--- B`, so `B` isn't dominated. In context of CFG I could imagine unreachable code blocks being a problem.

Reply:
Yes, the algorithm assumes that all vertices are reachable, since it relies on indexing the vertices when performing a DFS over the graph. The idea is to call it after `ControlFlowGraphBuilder::cleanUnreachable`, so we would be sure that there is no disconnected component.

Reply:
Right! In that case it's fine I suppose. Still it would be fairly easy to shrink down the dfs index vector to whatever index was actually used or have a more meaningful assertion (ie dfs index after dfs == expected n vertices) :)

Reply:
True. Yeah, maybe we could change that. The main reason I decided to go for static vector was to easily initialize the values of the auxiliaries variables used in the computation of the immediate dominator, see for example: https://github.com/ethereum/solidity/pull/14460/files#diff-e9e2916cc7d21e9eedfd8955be9a45fe9581588336f849f865993c76d530ae9dR182

Not really sure how to rewrite that logic, for instance: https://github.com/ethereum/solidity/pull/14460/files#diff-e9e2916cc7d21e9eedfd8955be9a45fe9581588336f849f865993c76d530ae9dR303 without having those default values already set.

Reply:
Can't you just assert [here](https://github.com/ethereum/solidity/pull/14460/files#diff-e9e2916cc7d21e9eedfd8955be9a45fe9581588336f849f865993c76d530ae9dR250) that `dfIdx` is `nVertices`?

Reply:
Yes, this we can do. I will push the change. But I don't think this would be sufficient to completely remove the necessity of knowing the number o vertices in advance.

Reply:
No it wouldn't but if you're assuming that the number of vertices is correct then this is a good place to test your assumption and things dont break further downstream b/c of it being violated :)

Reply:
> True. Yeah, maybe we could change that. The main reason I decided to go for static vector was to easily initialize the values of the auxiliaries variables used in the computation of the immediate dominator, see for example: https://github.com/ethereum/solidity/pull/14460/files#diff-e9e2916cc7d21e9eedfd8955be9a45fe9581588336f849f865993c76d530ae9dR182
> 
> Not really sure how to rewrite that logic, for instance: https://github.com/ethereum/solidity/pull/14460/files#diff-e9e2916cc7d21e9eedfd8955be9a45fe9581588336f849f865993c76d530ae9dR303 without having those default values already set.

you could grow the vectors as needed during dfs, no? but maybe I just don't see it right now

Reply:
> you could grow the vectors as needed during dfs, no? but maybe I just don't see it right now

Yes, but I would need to resize the vectors, at least the `parent` and `predecessors` at each iteration, which I would prefer to not do it. But yeah, I pushed a commit with the idea that you suggested. So we can think about which one we should go for ;)

Reply:
With the asserts in place that indeed the number of dfs indices coincides with the number of vertices provided to the constructor I think it's not necessary to dynamically grow the vectors. It does mean, that the size must be determined beforehand at some point, though. :)

Reply:
I think it's fine now, without the explicit size parameter. The algo will now simply process the only reachable connected component. If there are any others, they're no longer detectable, even with the assert for the size.

We could have kept the size arg just as a sanity check but I think it's not strictly necessary. I guess it's up to the caller now to ensure that there are no unreachable bits.

I actually added a note about that to the class docstring in one of the fixups. I also think this comment is now no longer true so I removed it:

> ```c++
> NOTE: The CFG can be disconnected, so the vertex IDs are not necessarily contiguous.
> ```

Reply:
Thanks for the comments and fixups @cameel, I applied them and squash/rebased with my changes based on Alex's comments. The comment still true though, since it refers to vertex IDs and not indices, but it is misplaced, it was supposed to be in the branch in which I use the dominator algorithm with the `CFG::BasicBlock`, not here.

Reply:
Oh. I thought it must be talking about DFS indices because the graph has no control over vertex IDs and treats them as opaque identifiers, not making assumptions about whether they are contiguous or not. So it wasn't making sense that way - IDs can always be non-contiguous even if the graph is connected. It really depends on the particular use of it.